stl/inc/xutility

```diff
@@ -4187,6 +4187,313 @@ struct default_sentinel_t {};
+    }
+
+    // clang-format off
+    constexpr counted_iterator operator++(int)
+        requires forward_iterator<_Iter>
+    {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Length > 0, "counted_iterator increment beyond end of range");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        counted_iterator _Tmp = *this;
+        ++_Current;
+        --_Length;
+        return _Tmp;
+    }
+
+    constexpr counted_iterator& operator--()
+        requires bidirectional_iterator<_Iter>
+    {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        if constexpr (_Offset_verifiable_v<_Iter>) {
+            _Current._Verify_offset(-1);
+        }
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        --_Current;
+        ++_Length;
+        return *this;
+    }
+
+    constexpr counted_iterator operator--(int)
+        requires bidirectional_iterator<_Iter>
+    {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        if constexpr (_Offset_verifiable_v<_Iter>) {
+            _Current._Verify_offset(-1);
+        }
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        counted_iterator _Tmp = *this;
+        --_Current;
+        ++_Length;
+        return _Tmp;
+    }
+
+    constexpr counted_iterator operator+(iter_difference_t<_Iter> _Diff) const
+        requires random_access_iterator<_Iter>
+    {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Diff <= _Length, "counted_iterator increment beyond end of range");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        return counted_iterator(_Current + _Diff, _Length - _Diff);
+    }
+
+    friend constexpr counted_iterator operator+(iter_difference_t<_Iter> _Diff, const counted_iterator& _Right)
+        requires random_access_iterator<_Iter>
+    {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Diff <= _Right.count(), "counted_iterator increment beyond end of range");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        return _Right + _Diff;
+    }
+
+    constexpr counted_iterator& operator+=(iter_difference_t<_Iter> _Diff)
+        requires random_access_iterator<_Iter>
+    {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Diff <= _Length, "counted_iterator increment beyond end of range");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        _Current += _Diff;
+        _Length -= _Diff;
+        return *this;
+    }
+
+    constexpr counted_iterator operator-(iter_difference_t<_Iter> _Diff) const
+        requires random_access_iterator<_Iter>
+    {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(-_Diff <= _Length, "counted_iterator decrement beyond end of range");
+        if constexpr (_Offset_verifiable_v<_Iter>) {
+            _Current._Verify_offset(-_Diff);
+        }
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        return counted_iterator(_Current - _Diff, _Length + _Diff);
+    }
+    // clang-format on
+
+    template <common_with<_Iter> _Other>
+    friend constexpr iter_difference_t<_Other> operator-(
+        const counted_iterator& _Left, const counted_iterator<_Other>& _Right) {
+        _Verify_Range(_Left, _Right);

```

Comment:
This requires `[_Left, _Right)` to be a valid range, but it could also be the case that `[_Right, Left)` is valid. Or it could be the case that neither of those is a valid range, but `_Left` and `_Right` both convert to `some_other_type` and `[some_other_type(_Left), some_other_type(_Right))` is a valid range. 

TLDR I don't think we can check easily here.

Reply:
That is not the case actually. The horror that is `iterators of the same range` in the context of `counted_iterator` is defined in the [standard](https://eel.is/c++draft/iterators.counted#counted.iterator-3).

I have no Idea why this is smashed in the actual synopsis but ...

Long story short this is responsible for the terrible thing I did with `_Verify_Range`. It is easy when we have the same iterator type then we can simply check `_STD next(_Left._Current, _Right._Length) == _STD next(_Left._Current, _Right._Length) `

My problem is what I should do with iterators of different type?

Reply:
Having now read the entire code (which I hadn't when I made this comment) I see you're doing something different than our conventional `_Verify_range`. I think that thing is necessary but not sufficient to implement the required semantics of `_Verify_range`, and should therefore have a different name like `_Same_sequence`. To be precise, I think `_Verify_range` needs to require `_Same_sequence(_Left, _Right) && _Left.count() >= _Right.count()`.

Reply:
I'm going to duplicate my latest comment into a new discussion so it doesn't get buried and resolve this discussion.